core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleSet.java

```diff
@@ -0,0 +1,101 @@
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package com.graphhopper.routing.util.spatialrules;
+
+import java.util.Collections;
+import java.util.List;
+
+import com.graphhopper.routing.profiles.RoadAccess;
+import com.graphhopper.routing.profiles.RoadClass;
+
+/**
+ * Contains all rules which are applicable for a certain position.
+ * 
+ * @author Thomas Butz
+ */
+public class SpatialRuleSet {
+    public static final SpatialRuleSet EMPTY = new SpatialRuleSet(Collections.<SpatialRule>emptyList(), -1);
+    
+    private final List<SpatialRule> rules;
+    private final int spatialId;
+
+    /**
+     * @param rules     a List of rules, ordered according to how they are to be applied
+     * @param spatialId the index of the rule with the highest priority
+     */
+    public SpatialRuleSet(List<SpatialRule> rules, int spatialId) {
+        this.rules = Collections.unmodifiableList(rules);
+        this.spatialId = spatialId;
+    }
+    
+    /**
+     * Return the max speed for a certain road class.
+     *
+     * @param roadClass       The highway type, e.g. {@link RoadClass#MOTORWAY}
+     * @param transport       The mode of transportation
+     * @param currentMaxSpeed The current max speed value or -1 if no value has been set yet
+     * @return the maximum speed value to be used
+     */
+    public double getMaxSpeed(RoadClass roadClass, TransportationMode transport, double currentMaxSpeed) {
+        double value = currentMaxSpeed;
+        for (SpatialRule rule : rules) {
+            value = rule.getMaxSpeed(roadClass, transport, value);
+        }
+        return value;
+    }
+
+    /**
+     * Returns the {@link RoadAccess} for a certain highway type and transportation mode.
+     *
+     * @param roadClass          The highway type, e.g. {@link RoadClass#MOTORWAY}
+     * @param transport          The mode of transportation
+     * @param currentRoadAccess  The current road access value (default: {@link RoadAccess#YES})
+     * @return the type of access to be used
+     */
+    public RoadAccess getAccess(RoadClass roadClass, TransportationMode transport, RoadAccess currentRoadAccess) {
+        RoadAccess value = currentRoadAccess;
+        for (SpatialRule rule : rules) {
+            value = rule.getAccess(roadClass, transport, value);
+        }
+        return value;
+    }
+    
+    /**
+     * @return the rules in this set
+     */
+    public List<SpatialRule> getRules() {
+        return rules;
+    }
+
+    /**
+     * @return the index of the rule with the highest priority or
+     *         <i>-1</i> if the set doesn't contain any rules
+     */
+    public int getSpatialId() {
+        return spatialId;
+    }
```

Comment:
Why do we need to add 1 when storing it into the edge? Also we should state it returns "EMPTY.getSpatialId" if it doesn't contain any rules.

Reply:
The spatial ID is stored as an IntEncodedValue and therefor needs to be unsigned. As -1 is used to indicate "no rules applied" i'm adding 1 to avoid signed values. But i'm open for suggestions :wink:

IMHO the Javadoc of SpatialRuleSet.getSpatialId() is pretty clear about this:
> the index of the rule with the highest priority or -1 if the set doesn't contain any rules

Reply:
Could we also use 0 for this "on rule" indicator? Like we currently seem to do:

```
spatialRuleEnc.setInt(false, edgeFlags, spatialRuleLookup.getSpatialId(rule));
```

Reply:
I think it's a matter of personal preference. -1 is a widely acknowledged default to indicate a missing element (e.g Javas List.indexOf()). The limitation to use non-negative numbers is enforced by the graph storage and should therefore be handled there.

Reply:
It is certainly a limitation of the storage and also the normal return value for a search is -1, but 0 is also the "Java default" for all numbers and "no rule" can be seen as the default value. Also I'm unsure why we should introduce this change when "empty" worked as 0 before.